File: algorithmSolution/leetCode/005_Longest Palindromic Substring.cpp
/*   DwyaneTalk@gmail.com
 *   by lutao
 *   leetCode : Longest Palindromic Substring
 *            : https://leetcode.com/problems/longest-palindromic-substring/
*/

#include<stdio.h>
// todo : Manacher method可以做到O(n)的时间复杂度
bool isPalindrome(char* s, int l, int r) {
    while(l < r) {
        if(s[l++] != s[r--])
            return false;
    }
    return true;
}

char* longestPalindrome(char* s) {
    int i = 1, max_l = 1, max_p = 0, left = 0;
    if(!s[0] || !s)    return s;
    while(s[i]) {
        if (i > max_l) {
            if(isPalindrome(s, i - max_l - 1, i)) {
                max_p = i - max_l - 1;
                max_l += 2;
            }
        }
        if(isPalindrome(s, i - max_l, i)) {
            max_p = i - max_l;
            max_l++;
        }
        i++;
    }
    s[max_p + max_l] = '\0';
    return s + max_p;
}

int main() {
    char s[] = "uuyyyu";
    printf("%s\n", s);
    char *ans = longestPalindrome(s);
    printf("%s\n", ans);
    return 0;
}

File: algorithmSolution/leetCodePython/005_Longest Palindromic Substring.py
'''
DwyaneTalk@gmail.com
https://leetcode.com/problems/two-sum/
'''
class Solution(object):
	def __init__(self):
		self.lo = 0
		self.maxlen = 0
	def longestPalindrome(self, s):
		ans = ""
		for i in range(len(s)):
			self.helper(s, i, i)
			self.helper(s, i, i + 1)
		return s[self.lo : self.lo + self.maxlen]

	def helper(self, s, lo, hi):
		sLen = len(s)
		while lo >=0 and hi < sLen and s[lo] == s[hi]:
			lo = lo - 1
			hi = hi + 1
		if hi - lo - 1 > self.maxlen:
			self.maxlen = hi - lo - 1
			self.lo = lo + 1

solution = Solution()
print(solution.longestPalindrome("bb"))